File: medium/Spiral_Matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        width = len(matrix[0])
        length = len(matrix)
        answer = []
        for r in range(int(min(width, length)/2+0.5)):
            x = r
            y = r
            answer.append(matrix[y][x])
            for w in range(1, width-2*r):
                x+=1
                answer.append(matrix[y][x])
            for l in range(1, length-2*r):
                y+=1
                answer.append(matrix[y][x])
            if length-2*r != 1:
                for w in range(1, width-2*r):
                    x-=1
                    answer.append(matrix[y][x])
            if width-2*r != 1:
                for l in range(1, length-2*r-1):
                    y-=1
                    answer.append(matrix[y][x])
        return answer

File: cpp/medium/Spiral_Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int length = matrix.size();
        int width = matrix[0].size();
        int count = length * width;
        vector<int> answer;
        int i = 0, j = 0, r = 0;
        while (true){
            while (i+r < width) {
                answer.push_back(matrix[j][i]);
                i++;
            }
            if (answer.size() == count) break;
            i--;
            j++;
            while (j+r < length) {
                answer.push_back(matrix[j][i]);
                j++;
            }
            if (answer.size() == count) break;
            j--;
            i--;
            while (i-r >= 0) {
                answer.push_back(matrix[j][i]);
                i--;
            }
            if (answer.size() == count) break;
            i++;
            j--;
            while (j-r >= 1) {
                answer.push_back(matrix[j][i]);
                j--;
            }
            if (answer.size() == count) break;
            i++;
            j++;
            r++;
        }
        return answer;
    }
};
